### ISSUE TITLE
VGG-16 Pretrained model did not increase the accuracy

### ISSUE THREAD

**Issue by sekti92** [keras-team/keras#12042]
Hello guys, I am trying to make a face recognition with VGG16 pretrained model in Keras.

However the training did not increase the accuracy.
I only have 46 data training, 26 validation, and 12 classes

How to increase the accuracy ? Thanks.

This is my actual code:
```
from keras.applications.vgg16 import VGG16
from keras.preprocessing import image
from keras.applications.vgg16 import preprocess_input
from keras.layers import Input, Flatten, Dense, Dropout
from keras.models import Model, Sequential
from keras import backend as K
import numpy as np
import matplotlib.pyplot as plt

# dimensions of our images.
from keras.preprocessing.image import ImageDataGenerator

img_width, img_height = 224, 224

train_data_dir = 'database/train'
validation_data_dir = 'database/validation'
nb_train_samples = 46
nb_validation_samples = 26
epochs = 50
batch_size = 2

if K.image_data_format() == 'channels_first':
    input_shape = (3, img_width, img_height)
else:
    input_shape = (img_width, img_height, 3)

#Get back the convolutional part of a VGG network trained on ImageNet
vgg_conv = VGG16(weights='imagenet', include_top=False, input_shape=input_shape)
vgg_conv.summary()
print('VGG Pretrained Model loaded.')

#Create your own input format (here 3x200x200)
model = Sequential()
model.add(vgg_conv)

#Add the fully-connected layers
model.add(Flatten())
model.add(Dense(4096, activation='relu', name='fc1'))
model.add(Dropout(0.3))
model.add(Dense(4096, activation='relu', name='fc2'))
model.add(Dropout(0.3))
model.add(Dense(12, activation='relu', name='prediction'))

#In the summary, weights and layers from VGG part will be hidden, but they will be fit during the training
model.summary()


# this is the augmentation configuration we will use for training
train_datagen = ImageDataGenerator(
    rescale=1. / 224,
    shear_range=0.2,
    zoom_range=0.2,
    horizontal_flip=True)

# this is the augmentation configuration we will use for testing:
# only rescaling
test_datagen = ImageDataGenerator(rescale=1. / 224)

train_generator = train_datagen.flow_from_directory(
    train_data_dir,
    target_size=(img_width, img_height),
    batch_size=batch_size,
    class_mode='categorical')

validation_generator = test_datagen.flow_from_directory(
    validation_data_dir,
    target_size=(img_width, img_height),
    batch_size=batch_size,
    class_mode='categorical')

# compile model
# model.compile(loss='sparse_categorical_crossentropy', optimizer=optimizers.RMSprop(lr=2e-4), metrics=['accuracy'])
model.compile(loss='categorical_crossentropy', optimizer='rmsprop', metrics=['accuracy'])

# Train the model
history = model.fit_generator(
    train_generator,
    steps_per_epoch=nb_train_samples / batch_size,
    epochs=epochs,
    validation_data=validation_generator,
    validation_steps=nb_validation_samples / batch_size)

# Save the model
model.save('vgg16_pretrained_2.h5')



# Check Performance
acc = history.history['acc']
val_acc = history.history['val_acc']
loss = history.history['loss']
val_loss = history.history['val_loss']

epochs = range(len(acc))

plt.plot(epochs, acc, 'b', label='Training acc')
plt.plot(epochs, val_acc, 'r', label='Validation acc')
plt.title('Training and validation accuracy')
plt.legend()

plt.figure()

plt.plot(epochs, loss, 'b', label='Training loss')
plt.plot(epochs, val_loss, 'r', label='Validation loss')
plt.title('Training and validation loss')
plt.legend()

plt.show()

```

On the 7 epoch the accuracy still the same:
```
Epoch 7/50
 1/23 [>.............................] - ETA: 50s - loss: 16.1181 - acc: 0.0000e+00
 2/23 [=>............................] - ETA: 49s - loss: 8.0590 - acc: 0.5000
 3/23 [==>...........................] - ETA: 46s - loss: 10.7454 - acc: 0.3333
 4/23 [====>.........................] - ETA: 44s - loss: 12.0886 - acc: 0.2500
 5/23 [=====>........................] - ETA: 42s - loss: 12.8945 - acc: 0.2000
 6/23 [======>.......................] - ETA: 40s - loss: 12.0886 - acc: 0.2500
 7/23 [========>.....................] - ETA: 37s - loss: 12.6642 - acc: 0.2143
 8/23 [=========>....................] - ETA: 35s - loss: 12.0886 - acc: 0.2500
 9/23 [==========>...................] - ETA: 33s - loss: 11.6408 - acc: 0.2778
10/23 [============>.................] - ETA: 30s - loss: 12.0886 - acc: 0.2500
11/23 [=============>................] - ETA: 28s - loss: 12.4549 - acc: 0.2273
12/23 [==============>...............] - ETA: 26s - loss: 12.7602 - acc: 0.2083
13/23 [===============>..............] - ETA: 23s - loss: 13.0185 - acc: 0.1923
14/23 [=================>............] - ETA: 21s - loss: 12.6642 - acc: 0.2143
15/23 [==================>...........] - ETA: 19s - loss: 12.8945 - acc: 0.2000
16/23 [===================>..........] - ETA: 16s - loss: 12.5923 - acc: 0.2188
17/23 [=====================>........] - ETA: 14s - loss: 12.7997 - acc: 0.2059
18/23 [======================>.......] - ETA: 11s - loss: 12.5363 - acc: 0.2222
19/23 [=======================>......] - ETA: 9s - loss: 12.7248 - acc: 0.2105
20/23 [=========================>....] - ETA: 7s - loss: 12.8945 - acc: 0.2000
21/23 [==========================>...] - ETA: 4s - loss: 13.0480 - acc: 0.1905
22/23 [===========================>..] - ETA: 2s - loss: 13.1875 - acc: 0.1818
23/23 [==============================] - 63s 3s/step - loss: 13.3149 - acc: 0.1739 - val_loss: 13.6384 - val_acc: 0.1538
```

**Comment by sekti92** [comment]
I try to change the epoch to 3. The loss become nan 

```
Epoch 30/50
 1/15 [>.............................] - ETA: 43s - loss: 10.7454 - acc: 0.3333
 2/15 [==>...........................] - ETA: 39s - loss: 8.0590 - acc: 0.5000
 3/15 [====>.........................] - ETA: 36s - loss: nan - acc: 0.4444
 4/15 [======>.......................] - ETA: 33s - loss: nan - acc: 0.3333
 5/15 [========>.....................] - ETA: 30s - loss: nan - acc: 0.2667
 6/15 [==========>...................] - ETA: 27s - loss: nan - acc: 0.2222
 7/15 [============>.................] - ETA: 24s - loss: nan - acc: 0.1905
 8/15 [==============>...............] - ETA: 21s - loss: nan - acc: 0.1667
 9/15 [================>.............] - ETA: 18s - loss: nan - acc: 0.1852
10/15 [==================>...........] - ETA: 15s - loss: nan - acc: 0.1667
11/15 [====================>.........] - ETA: 12s - loss: nan - acc: 0.1818
12/15 [======================>.......] - ETA: 9s - loss: nan - acc: 0.1944
13/15 [========================>.....] - ETA: 6s - loss: nan - acc: 0.2051
14/15 [==========================>...] - ETA: 3s - loss: nan - acc: 0.1905
15/15 [============================>.] - ETA: 0s - loss: nan - acc: 0.1778
16/15 [===============================] - 53s 3s/step - loss: nan - acc: 0.1880 - val_loss: nan - val_acc: 0.1154
```

**Comment by abtExp** [comment]
Use softmax activation.

**Comment by sekti92** [comment]
@abtExp It is works. But the accuracy did not increase:
```

 1/15 [>.............................] - ETA: 41s - loss: 10.7454 - acc: 0.3333
 2/15 [==>...........................] - ETA: 38s - loss: 10.7454 - acc: 0.3333
 3/15 [====>.........................] - ETA: 31s - loss: 12.5363 - acc: 0.2222
 4/15 [======>.......................] - ETA: 29s - loss: 13.4317 - acc: 0.1667
 5/15 [========>.....................] - ETA: 27s - loss: 12.8945 - acc: 0.2000
 6/15 [==========>...................] - ETA: 24s - loss: 13.4317 - acc: 0.1667
 7/15 [============>.................] - ETA: 22s - loss: 13.8155 - acc: 0.1429
 8/15 [==============>...............] - ETA: 20s - loss: 13.4317 - acc: 0.1667
 9/15 [================>.............] - ETA: 17s - loss: 13.7302 - acc: 0.1481
10/15 [==================>...........] - ETA: 14s - loss: 13.9690 - acc: 0.1333
11/15 [====================>.........] - ETA: 12s - loss: 14.1644 - acc: 0.1212
12/15 [======================>.......] - ETA: 9s - loss: 13.8795 - acc: 0.1389 
13/15 [========================>.....] - ETA: 6s - loss: 14.0517 - acc: 0.1282
14/15 [==========================>...] - ETA: 3s - loss: 14.1993 - acc: 0.1190
15/15 [============================>.] - ETA: 0s - loss: 14.3272 - acc: 0.1111
16/15 [===============================] - 51s 3s/step - loss: 14.4348 - acc: 0.1044 - val_loss: 14.2583 - val_acc: 0.1154

```

**Comment by colt18** [comment]
You need to elaborate the "face recognition" that you are dealing with. Is it 12 classes with total of 46 training and 26 validation images or for each class there are a total of 72 images. In either way you sample size is too small. In addition to this, CNN implementations that you use are better suited for large inter- class variation problems. For instance cat, dog and mouse classification would be more successeful then classifying 3 persons with different traits. In the end faces will look a like and your model may not found meaningful features to classify. BTW you val accuracy is just random guess. Try pre-deep learning methods for face recognition such as Eigenfaces, LBP etc.

**Comment by sekti92** [comment]
@colt18 hi thanks for the response. Sir now i am trying to change it using dlib library for face recognition

**Comment by msymp** [comment]
Thank you @colt18 . This issue is now closed. Thanks.

**Comment by qiaohong-li** [comment]
This is because you do not call the preprocess_input method coming with VGG net.
You scale the input by divide it by 224 to the range 0-1.
This is not keras VGG preprocessing way.
